### PR TITLE
Tensorflow local gradient aggregation.

### PR DESCRIPTION
**Pull request by bfrieds**
# Testing
I applied this patch to https://github.com/determined-ai/horovod/pull/10/commits/1889806baa120c15b5b9e5764d42a1a47168ccb4
```
diff --git a/horovod/tensorflow/__init__.py b/horovod/tensorflow/__init__.py
index 6632811..a38bd07 100644
--- a/horovod/tensorflow/__init__.py
+++ b/horovod/tensorflow/__init__.py
@@ -289,7 +289,7 @@ if _LegacyOptimizer is not None:
                     clear_op = grad_aggregator.assign(
                         grad_aggregator.initial_value)
                     clear_ops_list.append(clear_op)
-                return tf.group(*clear_ops_list)
+                with tf.control_dependencies([tf.print("CLEAR GRADS", self.counter)]):
+                     return tf.group(*clear_ops_list)

             def aggregate_grads():
@@ -297,9 +297,15 @@ if _LegacyOptimizer is not None:
                 if self._aggregation_frequency > 1:
                    for idx, grad in enumerate(self.grads):
                         grad_aggregation_variable_name = str(idx)
-                        grad_aggregator = tf.get_variable(
-                            grad_aggregation_variable_name, dtype=tf.float64)
-                        update_op = grad_aggregator.assign_add(grad)
+                        if idx == 0:
+                            print_op = tf.print("AGGREGATE GRADS", self.counter)
+                        else:
+                            print_op = tf.no_op()
+
+                        with tf.control_dependencies([print_op]):
+                            grad_aggregator = tf.get_variable(
+                                grad_aggregation_variable_name, dtype=tf.float64)
+                            update_op = grad_aggregator.assign_add(grad)
                         aggregation_ops_list.append(update_op)
                 return aggregation_ops_list

@@ -323,7 +329,7 @@ if _LegacyOptimizer is not None:
                     aggregated_grads = self.grads
                     aggregation_read_ops = tf.no_op()

-                with tf.control_dependencies([aggregation_read_ops]):
+                with tf.control_dependencies([aggregation_read_ops, tf.print("ALLREDUCE GRADS", self.counter)]):
                     averaged_gradients = self._allreduce_grads(self.grads)
                     with tf.control_dependencies([g.op for g in averaged_gradients]):
                         reset_op = self.counter.assign(
@@ -375,7 +381,7 @@ if _LegacyOptimizer is not None:
         def apply_gradients(self, *args, **kwargs):
             """Calls this same method on the underlying optimizer."""
             flattended_args0 = [item for tup in args[0] for item in tup]
-            with tf.control_dependencies([tf.group(*flattended_args0)]):
+            with tf.control_dependencies([tf.group(*flattended_args0), tf.print("APPLY GRADS", self.counter)]):
                 return tf.cond(tf.equal(self.counter, 0), lambda: self._optimizer.apply_gradients(*args, **kwargs), tf.no_op)

         def get_slot(self, *args, **kwargs):
```
and got the following expected output (excerpt) when using an aggregation frequency of 2 on 2 GPUs. Also confirmed that the final validation loss and accuracy on those two GPUs were equivalent.  (Horovod tensorflow_mnist_estimator.py example).
```
[1,1]<stderr>:AGGREGATE GRADS 0
[1,0]<stderr>:AGGREGATE GRADS 0
[1,0]<stderr>:APPLY GRADS 1
[1,0]<stderr>:AGGREGATE GRADS 1
[1,1]<stderr>:APPLY GRADS 1
[1,1]<stderr>:AGGREGATE GRADS 1
[1,0]<stderr>:ALLREDUCE GRADS 2
[1,1]<stderr>:ALLREDUCE GRADS 2
[1,0]<stderr>:APPLY GRADS 0
[1,1]<stderr>:APPLY GRADS 0
[1,1]<stderr>:CLEAR GRADS 0
[1,0]<stderr>:CLEAR GRADS 0
[1,1]<stderr>:AGGREGATE GRADS 0
[1,0]<stderr>:AGGREGATE GRADS 0
[1,0]<stderr>:APPLY GRADS 1
[1,1]<stderr>:APPLY GRADS 1
[1,0]<stderr>:AGGREGATE GRADS 1
[1,1]<stderr>:AGGREGATE GRADS 1
[1,0]<stderr>:ALLREDUCE GRADS 2
[1,1]<stderr>:ALLREDUCE GRADS 2
[1,0]<stderr>:APPLY GRADS 0
[1,1]<stderr>:APPLY GRADS 0
[1,1]<stderr>:CLEAR GRADS 0
[1,0]<stderr>:CLEAR GRADS 0
```

I tested "Refactor tf gradient aggregation." by running the tf estimator mnist example with aggregation frequency of 4 and verified that the model trained to a reasonable validation accuracy ~99% and that the validation metrics were the same on both GPUs.

### CODE REVIEW

File: horovod/tensorflow/__init__.py [hunk 1]
```diff
@@ -215,6 +215,123 @@ def allreduce_grads(grads):
...
+        with tf.variable_scope("aggregation_variables"):
+            self.counter = tf.get_variable(
+                "aggregation_counter", shape=(), dtype=tf.int32,
+                trainable=False, initializer=tf.zeros_initializer())
+            if self.aggregation_frequency > 1:
+                for idx, grad in enumerate(grads):
+                    grad_aggregation_variable_name = str(idx)
+                    grad_aggregation_variable = tf.get_variable(
+                        grad_aggregation_variable_name, shape=grad.get_shape().as_list(),
+                        trainable=False, initializer=tf.zeros_initializer(), dtype=tf.float64,
```
Comment (aaron276h) [inline horovod/tensorflow/__init__.py]: Same question as a couple other places about the `dtype`

File: horovod/tensorflow/__init__.py [hunk 2]
```diff
@@ -215,6 +215,123 @@ def allreduce_grads(grads):
...
+                        collections=[tf.GraphKeys.LOCAL_VARIABLES, "aggregating_collection"])
+                    self.gpu_shadow_vars.append(grad_aggregation_variable)
+                assert len(self.gpu_shadow_vars) == len(grads)
+
+    def clear_grads(self):
+        clear_ops_list = []
+        for idx, grad in enumerate(self.gpu_shadow_vars):
+            grad_aggregation_variable_name = str(idx)
+            grad_aggregator = tf.get_variable(
+                grad_aggregation_variable_name, dtype=tf.float64)
```
[the remark this answers is not in the document]
Comment (bfrieds) [reply horovod/tensorflow/__init__.py]: oof yeah -- so I got an error running the mnist example where it was using tf.float64 but the default is tf.float32

maybe what we need to do is make the `dtype` equal to whatever the `dtype` of `grad` is?

File: horovod/tensorflow/__init__.py [hunk 3]
```diff
@@ -215,6 +215,123 @@ def allreduce_grads(grads):
         return allreduce_grads
 
 
+class LocalGradientAggregationHelper:
+    def __init__(self, aggregation_frequency, allreduce_func=None):
+        if allreduce_func == None:
```
[the remark this answers is not in the document]
Comment (bfrieds) [reply horovod/tensorflow/__init__.py]: My idea was that we can turn this into something that is agnostic to whether or not we are in a distributed setting, with the goal of using this logic to support local gradient aggregation even in cases where `hvd.size() ==  1`.

File: horovod/tensorflow/__init__.py [hunk 4]
```diff
@@ -215,6 +215,123 @@ def allreduce_grads(grads):
         return allreduce_grads
 
 
+class LocalGradientAggregationHelper:
```
[the remark this answers is not in the document]
Comment (bfrieds) [reply horovod/tensorflow/__init__.py]: working towards it :) -- more work to be done around how we initialize variables, but planning on looking into that tomorrow

File: horovod/tensorflow/__init__.py [hunk 5]
```diff
@@ -252,15 +371,18 @@ def compute_gradients(self, *args, **kwargs):
...
+                allreduced_grads = self._agg_helper.compute_gradients(self.grads)
+                return list(zip(allreduced_grads, vars))
             else:
                 return gradients
 
         def apply_gradients(self, *args, **kwargs):
             """Calls this same method on the underlying optimizer."""
-            return self._optimizer.apply_gradients(*args, **kwargs)
+            flattended_args0 = [item for tup in args[0] for item in tup]
+            with tf.control_dependencies([tf.group(*flattended_args0)]):
```
[the remark this answers is not in the document]
Comment (bfrieds) [reply horovod/tensorflow/__init__.py]: Yeah I was thinking about that.. we have two ways I can think of

1) Give `LocalGradientAggregationHelper` access to the optimizer. But this might go counter to the goals I had of it being something that works for both TF Keras and Tensorflow. More exploration may be needed. But it would look something like:
```
def apply_gradients(self, optimizer, *args, **kwargs):
    flattended_args0 = [item for tup in args[0] for item in tup]
    with tf.control_dependencies([tf.group(*flattended_args0)]):
        return tf.cond(tf.equal(self.counter, 0), lambda: optimizer.apply_gradients(*args, **kwargs), tf.no_op)
```
2) Have a function that looks like 
```
def apply_gradients(self, apply_grads_func_closure, *args, **kwargs):
    flattended_args0 = [item for tup in args[0] for item in tup]
    with tf.control_dependencies([tf.group(*flattended_args0)]):
        return tf.cond(tf.equal(self.counter, 0), apply_grads_func_closure, tf.no_op)
```

thoughts?

File: horovod/tensorflow/__init__.py [hunk 6]
```diff
@@ -215,6 +215,123 @@ def allreduce_grads(grads):
...
+            clear_ops_list.append(clear_op)
+        return tf.group(*clear_ops_list)
+
+    def aggregate_grads(self, grads):
+        aggregation_ops_list = []
+        if self.aggregation_frequency > 1:
+            for idx, grad in enumerate(grads):
+                grad_aggregation_variable_name = str(idx)
+                grad_aggregator = tf.get_variable(
+                    grad_aggregation_variable_name, dtype=tf.float64)
```
[the remark this answers is not in the document]
Comment (bfrieds) [reply horovod/tensorflow/__init__.py]: I believe it fails, remember running into that

File: horovod/tensorflow/__init__.py [hunk 7]
```diff
@@ -215,6 +215,123 @@ def allreduce_grads(grads):
...
+            with tf.variable_scope("aggregation_variables", reuse=True):
+                clear_op = tf.cond(tf.equal(self.counter, 0), lambda: self.clear_grads(), tf.no_op)
+                with tf.control_dependencies([clear_op]):
+                    aggregation_ops_list = self.aggregate_grads(grads)
+
+                aggregation_ops = tf.group(*aggregation_ops_list)
+                with tf.control_dependencies([aggregation_ops]):
+                    update_ops = [self.counter.assign_add(tf.constant(1))]
+        else:
+            update_ops = [tf.no_op()]
```
[the remark this answers is not in the document]
Comment (bfrieds) [reply horovod/tensorflow/__init__.py]: these are not local_aggregation_ops though,  it's just updating the counter

File: horovod/tensorflow/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (aaron276h) [reply horovod/tensorflow/__init__.py]: Yeah we should definitely be using the `dtype` of `grad` if possible, and i don't see why it wouldn't be possible. Because it's possible we are communicating way too much if we are rounding everything up to `float64`. 

I am not opposed to using a dict, it's though considered better tensorflow practice to use `tf.get_variable`. I think we should be able to get rid of `dtype` in all the but the initial call which _should_ make this simpler.

File: horovod/tensorflow/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (aaron276h) [reply horovod/tensorflow/__init__.py]: Yeah I think second approach is better, as it's best to avoid passing `self._optimizer` into a function call.

File: horovod/tensorflow/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (aaron276h) [reply horovod/tensorflow/__init__.py]: I'd vote on taking this out if we aren't using it right now.

File: horovod/tensorflow/__init__.py [hunk 8]
```diff
@@ -215,6 +215,120 @@ def allreduce_grads(grads):
...
+                updated_grad_aggregator = grad_aggregator.assign_add(grad)
+                aggregation_ops_list.append(updated_grad_aggregator)
+        return aggregation_ops_list
+
+    def _allreduce_grads_helper(self, grads):
+        if self.aggregation_frequency > 1:
+            # Read in latest variables values.
+            aggregated_grads = []
+            aggregation_read_ops_list = []
+            with tf.variable_scope("aggregation_variables", reuse=True):
```
Comment (aaron276h) [inline horovod/tensorflow/__init__.py]: Don't think we need this anymore.

File: horovod/tensorflow/__init__.py [hunk 9]
```diff
@@ -215,6 +215,120 @@ def allreduce_grads(grads):
...
+                    return tuple(tf.divide(g, self.aggregation_frequency) for g in averaged_gradients)
+                else:
+                    # When grad updates are represented in `IndexedSlices`, we can not divide
+                    # them by int. Currently aggregation_frequency > 1 is not supported
+                    # `IndexedSlices`.
+                    return tuple(tf.identity(g) for g in averaged_gradients)
+
+    def compute_gradients(self, grads):
+        if self.aggregation_frequency > 1:
+            with tf.variable_scope("aggregation_variables", reuse=True):
```
Comment (aaron276h) [inline horovod/tensorflow/__init__.py]: Don't think we need this anymore either.

File: horovod/tensorflow/__init__.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (aaron276h) [reply horovod/tensorflow/__init__.py]: Would be nicer to get a more descriptive name here.
